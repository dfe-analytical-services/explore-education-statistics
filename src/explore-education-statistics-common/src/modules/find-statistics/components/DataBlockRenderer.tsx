import LoadingSpinner from '@common/components/LoadingSpinner';
import Tabs from '@common/components/Tabs';
import TabsSection from '@common/components/TabsSection';
import ChartRenderer from '@common/modules/charts/components/ChartRenderer';
import { GetInfographic } from '@common/modules/charts/components/InfographicBlock';
import { AxesConfiguration } from '@common/modules/charts/types/chart';
import getLabelDataSetConfigurations from '@common/modules/charts/util/getLabelDataSetConfigurations';
import useTableQuery from '@common/modules/find-statistics/hooks/useTableQuery';
import TimePeriodDataTable from '@common/modules/table-tool/components/TimePeriodDataTable';
import getDefaultTableHeaderConfig from '@common/modules/table-tool/utils/getDefaultTableHeadersConfig';
import mapTableHeadersConfig from '@common/modules/table-tool/utils/mapTableHeadersConfig';
import { DataBlock } from '@common/services/types/blocks';
import React, { MouseEvent, ReactNode } from 'react';

export interface DataBlockRendererProps {
  additionalTabContent?: ReactNode;
  dataBlock?: DataBlock;
  firstTabs?: ReactNode;
  lastTabs?: ReactNode;
  getInfographic?: GetInfographic;
  id: string;
  onToggle?: (section: { id: string; title: string }) => void;
  onSummaryDetailsToggle?: (
    isOpened: boolean,
    event: MouseEvent<HTMLElement>,
  ) => void;
}

const DataBlockRenderer = ({
  additionalTabContent,
  dataBlock,
  firstTabs,
  lastTabs,
  getInfographic,
  id,
  onToggle,
}: DataBlockRendererProps) => {
  const { value: fullTable, isLoading } = useTableQuery(
    dataBlock
      ? {
          ...dataBlock.dataBlockRequest,
          includeGeoJson: dataBlock.charts.some(chart => chart.type === 'map'),
        }
      : undefined,
  );

  return (
    <LoadingSpinner loading={isLoading}>
      <Tabs id={id} onToggle={onToggle}>
        {firstTabs}

        {dataBlock?.tables?.length && fullTable && (
          <TabsSection id={`${id}-tables`} title="Table">
            {dataBlock?.tables.map((table, index) => {
              return (
                <TimePeriodDataTable
                  key={index}
                  fullTable={fullTable}
                  captionTitle={dataBlock?.heading}
                  source={dataBlock?.source}
                  tableHeadersConfig={
                    table.tableHeaders
                      ? mapTableHeadersConfig(
                          table.tableHeaders,
                          fullTable.subjectMeta,
                        )
                      : getDefaultTableHeaderConfig(fullTable.subjectMeta)
                  }
                />
              );
            })}

            {additionalTabContent}
          </TabsSection>
        )}

        {dataBlock?.charts?.length && fullTable && (
          <TabsSection id={`${id}-charts`} title="Chart">
            <a
              className="govuk-visually-hidden"
              href={`#${id}-tables`}
              aria-live="assertive"
            >
              If you are using a keyboard or a screen reader you may wish to
              view the accessible table instead. Press enter to switch to the
              data tables tab.
            </a>

            {dataBlock?.charts.map((chart, index) => {
              const key = index;

              const axes = { ...chart.axes } as Required<AxesConfiguration>;

              if (
                axes.major?.dataSets?.some(dataSet => !dataSet.config) &&
                chart.labels
              ) {
                axes.major.dataSets = getLabelDataSetConfigurations(
                  chart.labels,
                  axes.major.dataSets,
                );
              }

              if (chart.type === 'infographic') {
                return (
                  <ChartRenderer
                    {...chart}
                    key={key}
                    axes={axes}
                    data={fullTable?.results}
                    meta={fullTable?.subjectMeta}
                    source={dataBlock?.source}
                    getInfographic={getInfographic}
                  />
                );
              }

              return (
                <ChartRenderer
                  {...chart}
                  key={key}
                  axes={axes}
                  data={fullTable?.results}
                  meta={fullTable?.subjectMeta}
                  source={dataBlock?.source}
                />
              );
            })}

            {additionalTabContent}
          </TabsSection>
        )}

        {lastTabs}
      </Tabs>
    </LoadingSpinner>
  );
};

export default DataBlockRenderer;
