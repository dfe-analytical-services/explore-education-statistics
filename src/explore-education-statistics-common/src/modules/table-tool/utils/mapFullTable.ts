import {
  CategoryFilter,
  Indicator,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import {
  FullTable,
  FullTableMeta,
} from '@common/modules/table-tool/types/fullTable';
import { TableDataResponse } from '@common/services/tableBuilderService';

export default function mapFullTable(
  unmappedFullTable: TableDataResponse,
): FullTable {
  const subjectMeta = unmappedFullTable.subjectMeta || {
    indicators: [],
    locations: [],
    timePeriodRange: [],
  };

  const filters = Object.values(unmappedFullTable.subjectMeta.filters).reduce<
    FullTableMeta['filters']
  >((acc, category) => {
    acc[category.legend] = {
      name: category.name,
      options: Object.values(category.options).flatMap(filterGroup =>
        filterGroup.options.map(
          option =>
            new CategoryFilter({
              ...option,
              group: filterGroup.label,
              category: category.legend,
              isTotal: category.totalValue === option.value,
            }),
        ),
      ),
    };

    return acc;
  }, {});

  return {
    ...unmappedFullTable,
    subjectMeta: {
      ...unmappedFullTable.subjectMeta,
      filters,
      indicators: subjectMeta.indicators.map(
        indicator => new Indicator(indicator),
      ),
      locations: subjectMeta.locations.map(
        location => new LocationFilter(location),
      ),
      timePeriodRange: subjectMeta.timePeriodRange.map(
        (timePeriod, order) => new TimePeriodFilter({ ...timePeriod, order }),
      ),
    },
  };
}
