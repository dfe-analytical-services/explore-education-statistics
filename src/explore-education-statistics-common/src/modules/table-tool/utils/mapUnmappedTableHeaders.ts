import {
  CategoryFilter,
  Filter,
  Indicator,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import { TableHeadersConfig } from '@common/modules/table-tool/types/tableHeaders';
import {
  TableHeader,
  UnmappedTableHeadersConfig,
} from '@common/services/permalinkService';

const mapToTableHeaders = (filters: Filter[]): TableHeader[] => {
  return filters.map<TableHeader>(filter => {
    if (filter instanceof LocationFilter) {
      return {
        type: 'Location',
        value: filter.value,
        level: filter.level,
      };
    }

    if (filter instanceof TimePeriodFilter) {
      return {
        type: 'TimePeriod',
        value: filter.value,
      };
    }

    if (filter instanceof CategoryFilter) {
      return {
        type: 'Filter',
        value: filter.value,
      };
    }

    if (filter instanceof Indicator) {
      return {
        type: 'Indicator',
        value: filter.value,
      };
    }

    throw new Error('Could not map filter to header');
  });
};

export default function mapUnmappedTableHeaders(
  tableHeaders: TableHeadersConfig,
): UnmappedTableHeadersConfig {
  return {
    columnGroups: tableHeaders.columnGroups.map(filters =>
      mapToTableHeaders(filters),
    ),
    columns: mapToTableHeaders(tableHeaders.columns),
    rowGroups: tableHeaders.rowGroups.map(filters =>
      mapToTableHeaders(filters),
    ),
    rows: mapToTableHeaders(tableHeaders.rows),
  };
}
