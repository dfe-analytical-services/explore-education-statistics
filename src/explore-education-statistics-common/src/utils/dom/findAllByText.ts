/**
 * Find elements that match the given
 * {@param text} and a {@param selector}.
 *
 * Additional options can be supplied via {@param options}.
 */
export default function findAllByText<T extends HTMLElement = HTMLElement>(
  text: string,
  selector: string,
  options?: {
    rootElement?: Element;
    useLowerCase?: boolean;
  },
): T[] {
  const { rootElement = document, useLowerCase = false } = options ?? {};

  return Array.from(rootElement.querySelectorAll(selector)).filter(element => {
    const searchText = useLowerCase
      ? (element.textContent || '').toLocaleLowerCase()
      : element.textContent || '';

    return searchText.includes(text);
  }) as T[];
}
