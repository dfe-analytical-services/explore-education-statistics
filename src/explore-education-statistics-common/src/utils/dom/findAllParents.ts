/**
 * Find all parent elements matching the given
 * {@param selector} for an {@param element}.
 */
export default function findAllParents<T extends HTMLElement = HTMLElement>(
  element: HTMLElement,
  selector: string,
): T[] {
  const parents: T[] = [];
  let currentElement = element.parentElement;

  while (currentElement && currentElement !== document.documentElement) {
    if (currentElement.matches(selector)) {
      parents.push(currentElement as T);
    }

    currentElement = currentElement.parentElement;
  }

  return parents;
}
