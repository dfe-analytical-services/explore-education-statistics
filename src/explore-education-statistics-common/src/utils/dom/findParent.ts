/**
 * Find a parent element matching the given
 * {@param selector} for an {@param element}.
 */
export default function findParent<T extends HTMLElement = HTMLElement>(
  element: HTMLElement,
  selector: string,
): T | null {
  let currentElement = element.parentElement;

  while (currentElement && currentElement !== document.documentElement) {
    if (currentElement.matches(selector)) {
      return currentElement as T;
    }

    currentElement = currentElement.parentElement;
  }

  return null;
}
