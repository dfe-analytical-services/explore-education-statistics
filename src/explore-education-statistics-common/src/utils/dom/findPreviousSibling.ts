/**
 * Find a sibling of an {@param element}
 * that matches the given {@param selector}.
 */
export default function findPreviousSibling<
  T extends HTMLElement = HTMLElement
>(element: Element, selector: string): T | null {
  let sibling = element.previousElementSibling;

  while (sibling && !sibling.matches(selector)) {
    sibling = sibling.previousElementSibling;
  }

  return (sibling as T) || null;
}
