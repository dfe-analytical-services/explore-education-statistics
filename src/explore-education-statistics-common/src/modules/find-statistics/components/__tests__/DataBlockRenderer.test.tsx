import {
  testChartConfiguration,
  testChartTableData,
  testDeprecatedChartConfiguration,
} from '@common/modules/charts/components/__tests__/__data__/testChartData';
import {
  testMapConfiguration,
  testMapTableData,
} from '@common/modules/charts/components/__tests__/__data__/testMapBlockData';
import getDefaultTableHeaderConfig from '@common/modules/table-tool/utils/getDefaultTableHeadersConfig';
import mapFullTable from '@common/modules/table-tool/utils/mapFullTable';
import mapUnmappedTableHeaders from '@common/modules/table-tool/utils/mapUnmappedTableHeaders';
import _tableBuilderService, {
  TableDataQuery,
} from '@common/services/tableBuilderService';
import {
  Chart,
  DataBlock,
  DataBlockRequest,
} from '@common/services/types/blocks';
import { waitFor, screen } from '@testing-library/dom';
import { render } from '@testing-library/react';
import { AxiosError } from 'axios';
import React from 'react';
import { forceVisible } from 'react-lazyload';
import DataBlockRenderer from '../DataBlockRenderer';

jest.mock('@common/services/tableBuilderService');

jest.mock('recharts/lib/util/LogUtils');

const tableBuilderService = _tableBuilderService as jest.Mocked<
  typeof _tableBuilderService
>;

describe('DataBlockRenderer', () => {
  const testDataBlock: DataBlock = {
    id: 'test-id',
    type: 'DataBlock',
    heading: '',
    order: 0,
    name: 'Test data block',
    source: '',
    dataBlockRequest: {
      subjectId: '1',
      geographicLevel: 'country',
      timePeriod: {
        startYear: 2012,
        startCode: 'AY',
        endYear: 2016,
        endCode: 'AY',
      },
      filters: ['characteristic-total', 'school-type-total'],
      indicators: [
        'authorised-absence-rate',
        'unauthorised-absence-rate',
        'overall-absence-rate',
      ],
      locations: {},
    },
    charts: [],
    tables: [],
  };

  test('renders error message if table response is error', async () => {
    tableBuilderService.getTableData.mockImplementation(() =>
      // eslint-disable-next-line prefer-promise-reject-errors
      Promise.reject({
        isAxiosError: true,
        message: 'Something went wrong',
        response: {
          status: 500,
        },
      } as AxiosError),
    );

    render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-datablock"
        dataBlock={{
          ...testDataBlock,
          charts: [testChartConfiguration],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(screen.getByText('Could not load content')).toBeInTheDocument();
    });
  });

  test('renders nothing if table response is 403', async () => {
    tableBuilderService.getTableData.mockImplementation(() =>
      // eslint-disable-next-line prefer-promise-reject-errors
      Promise.reject({
        isAxiosError: true,
        message: 'Forbidden',
        response: {
          status: 403,
        },
      } as AxiosError),
    );

    render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-datablock"
        dataBlock={{
          ...testDataBlock,
          charts: [testChartConfiguration],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(
        screen.queryByText('Could not load content'),
      ).not.toBeInTheDocument();
      expect(screen.queryByRole('table')).not.toBeInTheDocument();
    });
  });

  test('renders line chart', async () => {
    tableBuilderService.getTableData.mockImplementation(() =>
      Promise.resolve(testChartTableData),
    );

    const { container } = render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-datablock"
        dataBlock={{
          ...testDataBlock,
          charts: [testChartConfiguration],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(screen.getAllByRole('tab')).toHaveLength(1);

      expect(container.querySelectorAll('.recharts-line')).toHaveLength(3);
    });
  });

  test('renders horizontal chart', async () => {
    tableBuilderService.getTableData.mockImplementation(() =>
      Promise.resolve(testChartTableData),
    );

    const { container } = render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-block"
        dataBlock={{
          ...testDataBlock,
          charts: [
            {
              ...testChartConfiguration,
              type: 'horizontalbar',
            } as Chart,
          ],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(screen.getAllByRole('tab')).toHaveLength(1);
      expect(container.querySelectorAll('.recharts-bar')).toHaveLength(3);
    });
  });

  test('renders vertical chart', async () => {
    tableBuilderService.getTableData.mockImplementation(
      (_: DataBlockRequest, releaseId?: string) => {
        return Promise.resolve(testChartTableData);
      },
    );

    const { container } = render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-block"
        dataBlock={{
          ...testDataBlock,
          charts: [
            {
              ...testChartConfiguration,
              type: 'verticalbar',
            } as Chart,
          ],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(screen.getAllByRole('tab')).toHaveLength(1);
      expect(container.querySelectorAll('.recharts-bar')).toHaveLength(3);
    });
  });

  test('renders table', async () => {
    tableBuilderService.getTableData.mockImplementation(
      (_: DataBlockRequest, releaseId?: string) => {
        return Promise.resolve(testChartTableData);
      },
    );

    const fullTable = mapFullTable(testChartTableData);

    render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-block"
        dataBlock={{
          ...testDataBlock,
          tables: [
            {
              tableHeaders: mapUnmappedTableHeaders(
                getDefaultTableHeaderConfig(fullTable.subjectMeta),
              ),
              indicators: [
                'authorised-absence-rate',
                'unauthorised-absence-rate',
                'overall-absence-rate',
              ],
            },
          ],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(screen.getByRole('table')).toBeInTheDocument();
      expect(screen.getAllByRole('row')).toHaveLength(4);
      expect(screen.getAllByRole('cell')).toHaveLength(16);
    });
  });

  test('renders map', async () => {
    const getDataBlockForSubject = tableBuilderService.getTableData.mockImplementation(
      (_: DataBlockRequest, releaseId?: string) => {
        return Promise.resolve(testMapTableData);
      },
    );

    const { container } = render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-block"
        dataBlock={{
          ...testDataBlock,
          charts: [testMapConfiguration],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(getDataBlockForSubject).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: true,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(container.querySelector('.leaflet-container')).toBeInTheDocument();
    });
  });

  test('can render line chart with deprecated `labels` for data sets', async () => {
    tableBuilderService.getTableData.mockImplementation(() =>
      Promise.resolve(testChartTableData),
    );

    const { container } = render(
      <DataBlockRenderer
        releaseId="test-release-id"
        id="test-datablock"
        dataBlock={{
          ...testDataBlock,
          charts: [testDeprecatedChartConfiguration],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(tableBuilderService.getTableData).toBeCalledWith(
        {
          ...testDataBlock.dataBlockRequest,
          includeGeoJson: false,
        } as TableDataQuery,
        'test-release-id',
      );

      expect(
        container.querySelectorAll('section.govuk-tabs__panel'),
      ).toHaveLength(1);

      expect(container.querySelectorAll('.recharts-line')).toHaveLength(3);

      const legendItems = container.querySelectorAll('.recharts-legend-item');
      expect(legendItems[0]).toHaveTextContent(
        'Unauthorised absence rate (England)',
      );
      expect(legendItems[1]).toHaveTextContent(
        'Authorised absence rate (England)',
      );
      expect(legendItems[2]).toHaveTextContent(
        'Overall absence rate (England)',
      );
    });
  });

  test('re-rendering with new data block does not throw error', async () => {
    tableBuilderService.getTableData.mockImplementation(() =>
      Promise.resolve(testChartTableData),
    );

    const fullTable = mapFullTable(testChartTableData);

    const { rerender } = render(
      <DataBlockRenderer
        id="test-block"
        dataBlock={{
          ...testDataBlock,
          tables: [
            {
              tableHeaders: mapUnmappedTableHeaders(
                getDefaultTableHeaderConfig(fullTable.subjectMeta),
              ),
              indicators: [
                'authorised-absence-rate',
                'unauthorised-absence-rate',
                'overall-absence-rate',
              ],
            },
          ],
        }}
      />,
    );

    forceVisible();

    await waitFor(() => {
      expect(screen.getByRole('table')).toBeInTheDocument();
      expect(screen.getAllByRole('row')).toHaveLength(4);
      expect(screen.getAllByRole('cell')).toHaveLength(16);
    });

    tableBuilderService.getTableData.mockImplementation(() =>
      Promise.resolve({
        subjectMeta: {
          ...testChartTableData.subjectMeta,
          indicators: [
            {
              label: 'Number of authorised absence sessions',
              unit: '',
              value: 'authorised-absence-sessions',
              name: 'absence_sess',
            },
          ],
          timePeriodRange: [{ code: 'AY', label: '2018/19', year: 2018 }],
        },
        results: [
          {
            filters: ['characteristic-total', 'school-type-total'],
            geographicLevel: 'Country',
            location: { country: { code: 'E92000001', name: 'England' } },
            measures: {
              'authorised-absence-sessions': '500000',
            },
            timePeriod: '2018_AY',
          },
        ],
      }),
    );

    expect(() => {
      rerender(
        <DataBlockRenderer
          id="test-block"
          dataBlock={{
            ...testDataBlock,
            dataBlockRequest: {
              subjectId: '1',
              geographicLevel: 'country',
              timePeriod: {
                startYear: 2018,
                startCode: 'AY',
                endYear: 2018,
                endCode: 'AY',
              },
              filters: ['characteristic-total', 'school-type-total'],
              indicators: ['authorised-absence-sessions'],
              locations: {},
            },
            tables: [
              {
                tableHeaders: {
                  columnGroups: [],
                  rowGroups: [],
                  columns: [
                    {
                      type: 'TimePeriod',
                      value: '2018_AY',
                    },
                  ],
                  rows: [
                    {
                      type: 'Indicator',
                      value: 'authorised-absence-sessions',
                    },
                  ],
                },
                indicators: ['authorised-absence-sessions'],
              },
            ],
          }}
        />,
      );
    }).not.toThrowError();

    await waitFor(() => {
      expect(screen.getByRole('table')).toBeInTheDocument();
      expect(screen.getAllByRole('row')).toHaveLength(2);
      expect(screen.getAllByRole('cell')).toHaveLength(2);
    });
  });
});
