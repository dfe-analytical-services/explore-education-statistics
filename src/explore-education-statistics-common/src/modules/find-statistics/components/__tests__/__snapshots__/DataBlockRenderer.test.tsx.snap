// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataBlockRenderer renders map instead of chart 1`] = `
<section
  aria-labelledby="test-datablock-charts-tab"
  class="govuk-tabs__panel panel"
  id="test-datablock-charts"
  role="tabpanel"
  tabindex="-1"
>
  
  <a
    aria-live="assertive"
    class="govuk-visually-hidden"
    href="#test-datablock-tables"
  >
    If you are using a keyboard or a screen reader you may wish to view the accessible table instead. Press enter to switch to the data tables tab.
  </a>
  <figure
    class="govuk-!-margin-0"
  >
    <div
      class="govuk-!-margin-bottom-6"
    >
      <form>
        <div
          class="govuk-form-group"
        >
          <fieldset
            class="govuk-fieldset"
            id="data-and-location"
          >
            <legend
              class="govuk-fieldset__legend govuk-fieldset__legend--l govuk-visually-hidden"
            >
              Select data and location to view indicators
            </legend>
            <div
              class="mapContainer"
            >
              <div
                class="govuk-form-group mapIndicator"
              >
                <label
                  class="govuk-label"
                  for="selectedIndicator"
                >
                  1. Select data to view
                </label>
                <select
                  class="govuk-select govuk-!-width-full"
                  id="selectedIndicator"
                  name="selectedIndicator"
                >
                  <option
                    value="23_1_2_____"
                  >
                    Unauthorised absence rate
                  </option>
                  <option
                    value="26_1_2_____"
                  >
                    Overall absence rate
                  </option>
                  <option
                    value="28_1_2_____"
                  >
                    Authorised absence rate
                  </option>
                </select>
              </div>
              <div
                class="govuk-form-group mapLocation"
              >
                <label
                  class="govuk-label"
                  for="selectedLocation"
                >
                  2. Select a location
                </label>
                <select
                  class="govuk-select"
                  id="selectedLocation"
                  name="selectedLocation"
                >
                  <option
                    value=""
                  >
                    Select...
                  </option>
                  <option
                    value="E92000001"
                  >
                    England
                  </option>
                  <option
                    value="S92000001"
                  >
                    Scotland
                  </option>
                </select>
              </div>
            </div>
          </fieldset>
        </div>
      </form>
      <div
        class="mapContainer"
      >
        <div
          class="mapView"
        >
          <div
            class="map dfe-print-break-avoid leaflet-container leaflet-grab leaflet-touch-drag"
            style="width: 900px; height: 300px; position: relative;"
            tabindex="0"
          >
            <div
              class="leaflet-pane leaflet-map-pane"
              style="left: 0px; top: 0px;"
            >
              <div
                class="leaflet-pane leaflet-tile-pane"
              />
              <div
                class="leaflet-pane leaflet-shadow-pane"
              />
              <div
                class="leaflet-pane leaflet-overlay-pane"
              >
                <svg
                  height="0"
                  pointer-events="none"
                  style="left: 0px; top: 0px;"
                  viewBox="0 0 0 0"
                  width="0"
                >
                  <g>
                    <path
                      class="uk leaflet-interactive"
                      d="M-3 3L-3 -3L3 -3L3 3z"
                      fill="#3388ff"
                      fill-opacity="0.2"
                      fill-rule="evenodd"
                      stroke="#3388ff"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-opacity="1"
                      stroke-width="3"
                    />
                    <path
                      class="leaflet-interactive"
                      d="M0 0"
                      fill="#bf0000"
                      fill-opacity="0.2"
                      fill-rule="evenodd"
                      stroke="#3388ff"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-opacity="1"
                      stroke-width="3"
                    />
                  </g>
                </svg>
              </div>
              <div
                class="leaflet-pane leaflet-marker-pane"
              />
              <div
                class="leaflet-pane leaflet-tooltip-pane"
              />
              <div
                class="leaflet-pane leaflet-popup-pane"
              />
            </div>
            <div
              class="leaflet-control-container"
            >
              <div
                class="leaflet-top leaflet-left"
              >
                <div
                  class="leaflet-control-zoom leaflet-bar leaflet-control"
                >
                  <a
                    aria-label="Zoom in"
                    class="leaflet-control-zoom-in"
                    href="#"
                    role="button"
                    title="Zoom in"
                  >
                    +
                  </a>
                  <a
                    aria-label="Zoom out"
                    class="leaflet-control-zoom-out"
                    href="#"
                    role="button"
                    title="Zoom out"
                  >
                    −
                  </a>
                </div>
              </div>
              <div
                class="leaflet-top leaflet-right"
              />
              <div
                class="leaflet-bottom leaflet-left"
              />
              <div
                class="leaflet-bottom leaflet-right"
              >
                <div
                  class="leaflet-control-attribution leaflet-control"
                >
                  <a
                    href="https://leafletjs.com"
                    title="A JS library for interactive maps"
                  >
                    Leaflet
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-live="assertive"
          class="mapControls"
        >
          <div
            class="mapKey"
          >
            <h3
              class="govuk-heading-s"
            >
              Key to 
              Unauthorised absence rate
            </h3>
            <dl
              class="govuk-list"
            >
              <dd
                class="legend"
              >
                <span
                  class="rate0"
                  style="background-color: rgb(191, 0, 0);"
                >
                   
                </span>
                 -3% to -3%
              </dd>
              <dd
                class="legend"
              >
                <span
                  class="rate1"
                  style="background-color: rgb(143, 0, 0);"
                >
                   
                </span>
                 -3% to -3%
              </dd>
              <dd
                class="legend"
              >
                <span
                  class="rate2"
                  style="background-color: rgb(95, 0, 0);"
                >
                   
                </span>
                 -3% to -2%
              </dd>
              <dd
                class="legend"
              >
                <span
                  class="rate3"
                  style="background-color: rgb(47, 0, 0);"
                >
                   
                </span>
                 -2% to -2%
              </dd>
              <dd
                class="legend"
              >
                <span
                  class="rate4"
                  style="background-color: rgb(0, 0, 0);"
                >
                   
                </span>
                 -2% to -2%
              </dd>
            </dl>
          </div>
        </div>
      </div>
    </div>
    
  </figure>
</section>
`;

exports[`DataBlockRenderer renders table 1`] = `
<table
  aria-labelledby="dataTableCaption"
  class="govuk-table table"
>
  <thead>
    <tr>
      <td
        class="borderBottom"
        colspan="1"
        rowspan="1"
      />
      <th
        class="govuk-table__header--numeric borderBottom"
        colspan="1"
        rowspan="1"
        scope="col"
      >
        2014/15
      </th>
      <th
        class="govuk-table__header--numeric borderBottom borderRightNone"
        colspan="1"
        rowspan="1"
        scope="col"
      >
        2015/16
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th
        class=""
        colspan="1"
        rowspan="1"
        scope="row"
      >
        Unauthorised absence rate
      </th>
      <td
        class="govuk-table__cell--numeric borderBottom"
      >
        3%
      </td>
      <td
        class="govuk-table__cell--numeric borderBottom"
      >
        -3%
      </td>
    </tr>
  </tbody>
</table>
`;
